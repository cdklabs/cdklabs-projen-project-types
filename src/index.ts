export * from './cdk';
export * from './cdklabs/cdklabs';