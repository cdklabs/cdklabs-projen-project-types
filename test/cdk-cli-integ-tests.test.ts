import { Testing } from 'projen';
import { CdkCliIntegTestsWorkflow, yarn } from '../src';

test('snapshot test for the CDK CLI integ tests', () => {
  const repo = new yarn.CdkLabsMonorepo({
    name: 'monorepo',
    defaultReleaseBranch: 'main',
  });

  new CdkCliIntegTestsWorkflow(repo, {
    approvalEnvironment: 'approval',
    testEnvironment: 'test',
    buildRunsOn: 'runsOn',
    testRunsOn: 'testRunsOn',
    localPackages: ['@aws-cdk/bla', '@aws-cdk/bloeh'],
    sourceRepo: 'aws/some-repo',
    allowUpstreamVersions: ['@aws-cdk/bla'],
  });

  const outdir = Testing.synth(repo);
  expect(outdir).toMatchSnapshot();
});

test('snapshot test for the CDK CLI integ tests using atmosphere', () => {
  const repo = new yarn.CdkLabsMonorepo({
    name: 'monorepo',
    defaultReleaseBranch: 'main',
  });

  new CdkCliIntegTestsWorkflow(repo, {
    approvalEnvironment: 'approval',
    testEnvironment: 'test',
    buildRunsOn: 'runsOn',
    testRunsOn: 'testRunsOn',
    localPackages: ['@aws-cdk/bla', '@aws-cdk/bloeh'],
    sourceRepo: 'aws/some-repo',
    allowUpstreamVersions: ['@aws-cdk/bla'],
    enableAtmosphere: {
      endpoint: 'atmosphere.endpoint',
      pool: 'atmosphere.pool',
      oidcRoleArn: 'oidcRoleArn',
    },
  });

  const outdir = Testing.synth(repo);
  expect(outdir).toMatchSnapshot();
});

test('can set maxworkers', () => {
  const repo = new yarn.CdkLabsMonorepo({
    name: 'monorepo',
    defaultReleaseBranch: 'main',
  });

  new CdkCliIntegTestsWorkflow(repo, {
    approvalEnvironment: 'approval',
    testEnvironment: 'test',
    buildRunsOn: 'runsOn',
    testRunsOn: 'testRunsOn',
    localPackages: ['@aws-cdk/bla', '@aws-cdk/bloeh'],
    sourceRepo: 'aws/some-repo',
    maxWorkers: '500',
  });

  const outdir = Testing.synth(repo);
  expect(outdir['.github/workflows/integ.yml']).toContain('run: bin/run-suite --maxWorkers=500 --use-cli-release=');
});
