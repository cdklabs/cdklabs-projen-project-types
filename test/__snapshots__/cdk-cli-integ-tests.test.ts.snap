// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for the CDK CLI integ tests 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

* text=auto eol=lf
/.gitattributes linguist-generated
/.github/pull_request_template.md linguist-generated
/.github/workflows/auto-approve.yml linguist-generated
/.github/workflows/auto-queue.yml linguist-generated
/.github/workflows/build.yml linguist-generated
/.github/workflows/integ.yml linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.github/workflows/upgrade.yml linguist-generated
/.gitignore linguist-generated
/.npmignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/LICENSE linguist-generated
/package.json linguist-generated
/tsconfig.dev.json linguist-generated
/tsconfig.json linguist-generated
/yarn.lock linguist-generated",
  ".github/pull_request_template.md": "Fixes #",
  ".github/workflows/auto-approve.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: auto-approve
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
jobs:
  approve:
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: contains(github.event.pull_request.labels.*.name, 'auto-approve') && (github.event.pull_request.user.login == 'github-bot' || github.event.pull_request.user.login == 'cdklabs-automation' || github.event.pull_request.user.login == 'dependabot[bot]')
    steps:
      - uses: hmarr/auto-approve-action@v2.2.1
        with:
          github-token: \${{ secrets.GITHUB_TOKEN }}
",
  ".github/workflows/auto-queue.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: auto-queue
on:
  pull_request_target:
    types:
      - opened
      - reopened
      - ready_for_review
jobs:
  enableAutoQueue:
    name: "Set AutoQueue on PR #\${{ github.event.number }}"
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
      contents: write
    steps:
      - uses: peter-evans/enable-pull-request-automerge@v3
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          pull-request-number: \${{ github.event.number }}
          merge-method: squash
",
  ".github/workflows/build.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: build
on:
  pull_request: {}
  workflow_dispatch: {}
  merge_group: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    outputs:
      self_mutation_happened: \${{ steps.self_mutation.outputs.self_mutation_happened }}
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: lts/*
      - name: Install dependencies
        run: yarn install --check-files
      - name: build
        run: npx projen build
      - name: Find mutations
        id: self_mutation
        run: |-
          git add .
          git diff --staged --patch --exit-code > repo.patch || echo "self_mutation_happened=true" >> $GITHUB_OUTPUT
        working-directory: ./
      - name: Upload patch
        if: steps.self_mutation.outputs.self_mutation_happened
        uses: actions/upload-artifact@v4.4.0
        with:
          name: repo.patch
          path: repo.patch
          overwrite: true
      - name: Fail build on mutation
        if: steps.self_mutation.outputs.self_mutation_happened
        run: |-
          echo "::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch."
          cat repo.patch
          exit 1
  self-mutation:
    needs: build
    runs-on: ubuntu-latest
    permissions:
      contents: write
    if: always() && needs.build.outputs.self_mutation_happened && !(github.event.pull_request.head.repo.full_name != github.repository)
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Download patch
        uses: actions/download-artifact@v4
        with:
          name: repo.patch
          path: \${{ runner.temp }}
      - name: Apply patch
        run: '[ -s \${{ runner.temp }}/repo.patch ] && git apply \${{ runner.temp }}/repo.patch || echo "Empty patch. Skipping."'
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: Push changes
        env:
          PULL_REQUEST_REF: \${{ github.event.pull_request.head.ref }}
        run: |-
          git add .
          git commit -s -m "chore: self mutation"
          git push origin HEAD:$PULL_REQUEST_REF
",
  ".github/workflows/integ.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: integ
on:
  pull_request_target:
    branches:
      - main
jobs:
  prepare:
    runs-on: runsOn
    permissions:
      contents: read
    environment: approval
    env:
      CI: "true"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ github.event.pull_request.head.sha }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
          fetch-depth: 0
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: lts/*
      - name: Install dependencies
        run: yarn install --check-files
      - name: Bump to realistic versions
        env:
          TESTING_CANDIDATE: "true"
        run: yarn workspaces run bump
      - name: build
        env:
          RELEASE: "true"
        run: npx projen build
      - name: Upload artifact
        uses: actions/upload-artifact@v4.4.0
        with:
          name: build-artifact
          path: packages/**/dist/js/*.tgz
          overwrite: "true"
  integ_matrix:
    needs: prepare
    runs-on: testRunsOn
    permissions:
      contents: read
      id-token: write
    environment: test
    env:
      MAVEN_ARGS: --no-transfer-progress
      IS_CANARY: "true"
      CI: "true"
    steps:
      - name: Download build artifacts
        uses: actions/download-artifact@v4
        with:
          name: build-artifact
          path: packages
      - name: Set up JDK 18
        if: matrix.suite == 'init-java' || matrix.suite == 'cli-integ-tests'
        uses: actions/setup-java@v4
        with:
          java-version: "18"
          distribution: corretto
      - name: Authenticate Via OIDC Role
        id: creds
        uses: aws-actions/configure-aws-credentials@v4
        with:
          aws-region: us-east-1
          role-duration-seconds: 14400
          role-to-assume: \${{ vars.AWS_ROLE_TO_ASSUME_FOR_TESTING }}
          role-session-name: run-tests@aws-cdk-cli-integ
          output-credentials: true
      - name: Set git identity
        run: |-
          git config --global user.name "aws-cdk-cli-integ"
          git config --global user.email "noreply@example.com"
      - name: Install Verdaccio
        run: npm install -g verdaccio pm2
      - name: Create Verdaccio config
        run: |-
          mkdir -p $HOME/.config/verdaccio
          echo '{"storage":"./storage","auth":{"htpasswd":{"file":"./htpasswd"}},"uplinks":{"npmjs":{"url":"https://registry.npmjs.org/"}},"packages":{"@aws-cdk/bla":{"access":"$all","publish":"$all","proxy":"none"},"@aws-cdk/bloeh":{"access":"$all","publish":"$all","proxy":"none"},"**":{"access":"$all","proxy":"npmjs"}}}' > $HOME/.config/verdaccio/config.yaml
      - name: Start Verdaccio
        run: |-
          pm2 start verdaccio -- --config $HOME/.config/verdaccio/config.yaml
          sleep 5 # Wait for Verdaccio to start
      - name: Configure npm to use local registry
        run: |-
          echo '//localhost:4873/:_authToken="MWRjNDU3OTE1NTljYWUyOTFkMWJkOGUyYTIwZWMwNTI6YTgwZjkyNDE0NzgwYWQzNQ=="' > ~/.npmrc
          echo 'registry=http://localhost:4873/' >> ~/.npmrc
      - name: Find an locally publish all tarballs
        run: |-
          for pkg in packages/{@aws-cdk/bla,@aws-cdk/bloeh}/dist/js/*.tgz; do
            npm publish $pkg
          done
      - name: Download and install the test artifact
        run: |-
          npm install @aws-cdk-testing/cli-integ
          mv ./node_modules/@aws-cdk-testing/cli-integ/* .
      - name: Determine latest package versions
        id: versions
        run: |-
          CLI_VERSION=$(cd \${TMPDIR:-/tmp} && npm view aws-cdk version)
          echo "CLI version: \${CLI_VERSION}"
          echo "cli_version=\${CLI_VERSION}" >> $GITHUB_OUTPUT
          LIB_VERSION=$(cd \${TMPDIR:-/tmp} && npm view aws-cdk-lib version)
          echo "lib version: \${LIB_VERSION}"
          echo "lib_version=\${LIB_VERSION}" >> $GITHUB_OUTPUT
      - name: "Run the test suite: \${{ matrix.suite }}"
        env:
          JEST_TEST_CONCURRENT: \${{ matrix.suite == 'cli-integ-tests' && 'true' || 'false' }}
          JSII_SILENCE_WARNING_DEPRECATED_NODE_VERSION: "true"
          JSII_SILENCE_WARNING_UNTESTED_NODE_VERSION: "true"
          JSII_SILENCE_WARNING_KNOWN_BROKEN_NODE_VERSION: "true"
          DOCKERHUB_DISABLED: "true"
          AWS_REGIONS: us-east-2,eu-west-1,eu-north-1,ap-northeast-1,ap-south-1
          CDK_MAJOR_VERSION: "2"
          RELEASE_TAG: latest
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        run: bin/run-suite --use-cli-release=\${{ steps.versions.outputs.cli_version }} --framework-version=\${{ steps.versions.outputs.lib_version }} \${{ matrix.suite }}
    strategy:
      fail-fast: false
      matrix:
        suite:
          - cli-integ-tests
          - init-csharp
          - init-fsharp
          - init-go
          - init-java
          - init-javascript
          - init-python
          - init-typescript-app
          - init-typescript-lib
          - tool-integrations
  integ:
    needs: integ_matrix
    runs-on: testRunsOn
    permissions: {}
    if: always()
    steps:
      - name: Build result
        run: echo \${{ needs.integ_matrix.result }}
      - name: Set status based on matrix build
        if: \${{ needs.integ_matrix.result != 'success' }}
        run: exit 1
",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".github/workflows/upgrade.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: upgrade
on:
  workflow_dispatch: {}
  schedule:
    - cron: 0 0 * * *
jobs:
  upgrade:
    name: Upgrade
    runs-on: ubuntu-latest
    permissions:
      contents: read
    outputs:
      patch_created: \${{ steps.create_patch.outputs.patch_created }}
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: lts/*
      - name: Install dependencies
        run: yarn install --check-files --frozen-lockfile
      - name: Upgrade dependencies
        run: npx projen upgrade
      - name: Find mutations
        id: create_patch
        run: |-
          git add .
          git diff --staged --patch --exit-code > repo.patch || echo "patch_created=true" >> $GITHUB_OUTPUT
        working-directory: ./
      - name: Upload patch
        if: steps.create_patch.outputs.patch_created
        uses: actions/upload-artifact@v4.4.0
        with:
          name: repo.patch
          path: repo.patch
          overwrite: true
  pr:
    name: Create Pull Request
    needs: upgrade
    runs-on: ubuntu-latest
    permissions:
      contents: read
    if: \${{ needs.upgrade.outputs.patch_created }}
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Download patch
        uses: actions/download-artifact@v4
        with:
          name: repo.patch
          path: \${{ runner.temp }}
      - name: Apply patch
        run: '[ -s \${{ runner.temp }}/repo.patch ] && git apply \${{ runner.temp }}/repo.patch || echo "Empty patch. Skipping."'
      - name: Set git identity
        run: |-
          git config user.name "github-actions"
          git config user.email "github-actions@github.com"
      - name: Create Pull Request
        id: create-pr
        uses: peter-evans/create-pull-request@v6
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          commit-message: |-
            chore(deps): upgrade dependencies

            Upgrades project dependencies. See details in [workflow run].

            [Workflow Run]: \${{ github.server_url }}/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

            ------

            *Automatically created by projen via the "upgrade" workflow*
          branch: github-actions/upgrade
          title: "chore(deps): upgrade dependencies"
          labels: auto-approve
          body: |-
            Upgrades project dependencies. See details in [workflow run].

            [Workflow Run]: \${{ github.server_url }}/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

            ------

            *Automatically created by projen via the "upgrade" workflow*
          author: github-actions <github-actions@github.com>
          committer: github-actions <github-actions@github.com>
          signoff: true
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/workflows/auto-approve.yml
!/package.json
!/LICENSE
!/.npmignore
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
pids
*.pid
*.seed
*.pid.lock
lib-cov
coverage
*.lcov
.nyc_output
build/Release
node_modules/
jspm_packages/
*.tsbuildinfo
.eslintcache
*.tgz
.yarn-integrity
.cache
!/.github/workflows/build.yml
!/.github/workflows/upgrade.yml
!/.github/pull_request_template.md
!/test/
!/tsconfig.json
!/tsconfig.dev.json
!/src/
/lib
/dist/
!/.github/workflows/auto-queue.yml
!/.github/workflows/integ.yml
!/.projenrc.ts
",
  ".npmignore": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
/.projen/
permissions-backup.acl
/test/
/tsconfig.dev.json
/src/
!/lib/
!/lib/**/*.js
!/lib/**/*.d.ts
dist
/tsconfig.json
/.github/
/.vscode/
/.idea/
/.projenrc.js
tsconfig.tsbuildinfo
/.gitattributes
/.projenrc.ts
/projenrc
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "dependencies": [
      {
        "name": "@types/node",
        "type": "build",
      },
      {
        "name": "constructs",
        "type": "build",
        "version": "^10.0.0",
      },
      {
        "name": "projen",
        "type": "build",
      },
      {
        "name": "ts-node",
        "type": "build",
      },
      {
        "name": "typescript",
        "type": "build",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "files": [
      ".gitattributes",
      ".github/pull_request_template.md",
      ".github/workflows/auto-approve.yml",
      ".github/workflows/auto-queue.yml",
      ".github/workflows/build.yml",
      ".github/workflows/integ.yml",
      ".github/workflows/pull-request-lint.yml",
      ".github/workflows/upgrade.yml",
      ".gitignore",
      ".npmignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "LICENSE",
      "tsconfig.dev.json",
      "tsconfig.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "env": {
      "PATH": "$(npx -c "node --print process.env.PATH")",
    },
    "tasks": {
      "build": {
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "exec": "yarn workspaces run build",
          },
        ],
      },
      "clobber": {
        "condition": "git diff --exit-code > /dev/null",
        "description": "hard resets to HEAD of origin and cleans the local repo",
        "env": {
          "BRANCH": "$(git branch --show-current)",
        },
        "name": "clobber",
        "steps": [
          {
            "exec": "git checkout -b scratch",
            "name": "save current HEAD in "scratch" branch",
          },
          {
            "exec": "git checkout $BRANCH",
          },
          {
            "exec": "git fetch origin",
            "name": "fetch latest changes from origin",
          },
          {
            "exec": "git reset --hard origin/$BRANCH",
            "name": "hard reset to origin commit",
          },
          {
            "exec": "git clean -fdx",
            "name": "clean all untracked files",
          },
          {
            "say": "ready to rock! (unpushed commits are under the "scratch" branch)",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
        "steps": [
          {
            "exec": "yarn workspaces run compile",
          },
        ],
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
        "steps": [
          {
            "exec": "ts-node --project tsconfig.dev.json .projenrc.ts",
          },
        ],
      },
      "install": {
        "description": "Install project dependencies and update lockfile (non-frozen)",
        "name": "install",
        "steps": [
          {
            "exec": "yarn install --check-files",
          },
        ],
      },
      "install:ci": {
        "description": "Install project dependencies using frozen lockfile",
        "name": "install:ci",
        "steps": [
          {
            "exec": "yarn install --check-files --frozen-lockfile",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
        "steps": [
          {
            "exec": "yarn workspaces run package",
          },
        ],
      },
      "post-upgrade": {
        "description": "Runs after upgrading dependencies",
        "name": "post-upgrade",
      },
      "run": {
        "name": "run",
        "steps": [
          {
            "exec": "yarn workspaces run",
            "receiveArgs": true,
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
        "steps": [
          {
            "exec": "yarn workspaces run test",
          },
        ],
      },
      "upgrade": {
        "description": "Upgrade dependencies in all workspaces",
        "env": {
          "CI": "0",
        },
        "name": "upgrade",
        "steps": [
          {
            "exec": "npx npm-check-updates@16 --dep=dev,optional,peer,prod,bundle --upgrade --target=minor",
          },
          {
            "exec": "yarn workspaces run check-for-updates",
          },
          {
            "exec": "yarn install --check-files",
          },
          {
            "exec": "yarn upgrade",
          },
          {
            "spawn": "default",
          },
          {
            "spawn": "post-upgrade",
          },
        ],
      },
    },
  },
  "LICENSE": "
                                 Apache License
                           Version 2.0, January 2004
                        http://www.apache.org/licenses/

   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION

   1. Definitions.

      "License" shall mean the terms and conditions for use, reproduction,
      and distribution as defined by Sections 1 through 9 of this document.

      "Licensor" shall mean the copyright owner or entity authorized by
      the copyright owner that is granting the License.

      "Legal Entity" shall mean the union of the acting entity and all
      other entities that control, are controlled by, or are under common
      control with that entity. For the purposes of this definition,
      "control" means (i) the power, direct or indirect, to cause the
      direction or management of such entity, whether by contract or
      otherwise, or (ii) ownership of fifty percent (50%) or more of the
      outstanding shares, or (iii) beneficial ownership of such entity.

      "You" (or "Your") shall mean an individual or Legal Entity
      exercising permissions granted by this License.

      "Source" form shall mean the preferred form for making modifications,
      including but not limited to software source code, documentation
      source, and configuration files.

      "Object" form shall mean any form resulting from mechanical
      transformation or translation of a Source form, including but
      not limited to compiled object code, generated documentation,
      and conversions to other media types.

      "Work" shall mean the work of authorship, whether in Source or
      Object form, made available under the License, as indicated by a
      copyright notice that is included in or attached to the work
      (an example is provided in the Appendix below).

      "Derivative Works" shall mean any work, whether in Source or Object
      form, that is based on (or derived from) the Work and for which the
      editorial revisions, annotations, elaborations, or other modifications
      represent, as a whole, an original work of authorship. For the purposes
      of this License, Derivative Works shall not include works that remain
      separable from, or merely link (or bind by name) to the interfaces of,
      the Work and Derivative Works thereof.

      "Contribution" shall mean any work of authorship, including
      the original version of the Work and any modifications or additions
      to that Work or Derivative Works thereof, that is intentionally
      submitted to Licensor for inclusion in the Work by the copyright owner
      or by an individual or Legal Entity authorized to submit on behalf of
      the copyright owner. For the purposes of this definition, "submitted"
      means any form of electronic, verbal, or written communication sent
      to the Licensor or its representatives, including but not limited to
      communication on electronic mailing lists, source code control systems,
      and issue tracking systems that are managed by, or on behalf of, the
      Licensor for the purpose of discussing and improving the Work, but
      excluding communication that is conspicuously marked or otherwise
      designated in writing by the copyright owner as "Not a Contribution."

      "Contributor" shall mean Licensor and any individual or Legal Entity
      on behalf of whom a Contribution has been received by Licensor and
      subsequently incorporated within the Work.

   2. Grant of Copyright License. Subject to the terms and conditions of
      this License, each Contributor hereby grants to You a perpetual,
      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
      copyright license to reproduce, prepare Derivative Works of,
      publicly display, publicly perform, sublicense, and distribute the
      Work and such Derivative Works in Source or Object form.

   3. Grant of Patent License. Subject to the terms and conditions of
      this License, each Contributor hereby grants to You a perpetual,
      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
      (except as stated in this section) patent license to make, have made,
      use, offer to sell, sell, import, and otherwise transfer the Work,
      where such license applies only to those patent claims licensable
      by such Contributor that are necessarily infringed by their
      Contribution(s) alone or by combination of their Contribution(s)
      with the Work to which such Contribution(s) was submitted. If You
      institute patent litigation against any entity (including a
      cross-claim or counterclaim in a lawsuit) alleging that the Work
      or a Contribution incorporated within the Work constitutes direct
      or contributory patent infringement, then any patent licenses
      granted to You under this License for that Work shall terminate
      as of the date such litigation is filed.

   4. Redistribution. You may reproduce and distribute copies of the
      Work or Derivative Works thereof in any medium, with or without
      modifications, and in Source or Object form, provided that You
      meet the following conditions:

      (a) You must give any other recipients of the Work or
          Derivative Works a copy of this License; and

      (b) You must cause any modified files to carry prominent notices
          stating that You changed the files; and

      (c) You must retain, in the Source form of any Derivative Works
          that You distribute, all copyright, patent, trademark, and
          attribution notices from the Source form of the Work,
          excluding those notices that do not pertain to any part of
          the Derivative Works; and

      (d) If the Work includes a "NOTICE" text file as part of its
          distribution, then any Derivative Works that You distribute must
          include a readable copy of the attribution notices contained
          within such NOTICE file, excluding those notices that do not
          pertain to any part of the Derivative Works, in at least one
          of the following places: within a NOTICE text file distributed
          as part of the Derivative Works; within the Source form or
          documentation, if provided along with the Derivative Works; or,
          within a display generated by the Derivative Works, if and
          wherever such third-party notices normally appear. The contents
          of the NOTICE file are for informational purposes only and
          do not modify the License. You may add Your own attribution
          notices within Derivative Works that You distribute, alongside
          or as an addendum to the NOTICE text from the Work, provided
          that such additional attribution notices cannot be construed
          as modifying the License.

      You may add Your own copyright statement to Your modifications and
      may provide additional or different license terms and conditions
      for use, reproduction, or distribution of Your modifications, or
      for any such Derivative Works as a whole, provided Your use,
      reproduction, and distribution of the Work otherwise complies with
      the conditions stated in this License.

   5. Submission of Contributions. Unless You explicitly state otherwise,
      any Contribution intentionally submitted for inclusion in the Work
      by You to the Licensor shall be under the terms and conditions of
      this License, without any additional terms or conditions.
      Notwithstanding the above, nothing herein shall supersede or modify
      the terms of any separate license agreement you may have executed
      with Licensor regarding such Contributions.

   6. Trademarks. This License does not grant permission to use the trade
      names, trademarks, service marks, or product names of the Licensor,
      except as required for reasonable and customary use in describing the
      origin of the Work and reproducing the content of the NOTICE file.

   7. Disclaimer of Warranty. Unless required by applicable law or
      agreed to in writing, Licensor provides the Work (and each
      Contributor provides its Contributions) on an "AS IS" BASIS,
      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
      implied, including, without limitation, any warranties or conditions
      of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A
      PARTICULAR PURPOSE. You are solely responsible for determining the
      appropriateness of using or redistributing the Work and assume any
      risks associated with Your exercise of permissions under this License.

   8. Limitation of Liability. In no event and under no legal theory,
      whether in tort (including negligence), contract, or otherwise,
      unless required by applicable law (such as deliberate and grossly
      negligent acts) or agreed to in writing, shall any Contributor be
      liable to You for damages, including any direct, indirect, special,
      incidental, or consequential damages of any character arising as a
      result of this License or out of the use or inability to use the
      Work (including but not limited to damages for loss of goodwill,
      work stoppage, computer failure or malfunction, or any and all
      other commercial damages or losses), even if such Contributor
      has been advised of the possibility of such damages.

   9. Accepting Warranty or Additional Liability. While redistributing
      the Work or Derivative Works thereof, You may choose to offer,
      and charge a fee for, acceptance of support, warranty, indemnity,
      or other liability obligations and/or rights consistent with this
      License. However, in accepting such obligations, You may act only
      on Your own behalf and on Your sole responsibility, not on behalf
      of any other Contributor, and only if You agree to indemnify,
      defend, and hold each Contributor harmless for any liability
      incurred by, or claims asserted against, such Contributor by reason
      of your accepting any such warranty or additional liability.

   END OF TERMS AND CONDITIONS

   APPENDIX: How to apply the Apache License to your work.

      To apply the Apache License to your work, attach the following
      boilerplate notice, with the fields enclosed by brackets "[]"
      replaced with your own identifying information. (Don't include
      the brackets!)  The text should be enclosed in the appropriate
      comment syntax for the file format. We also recommend that a
      file or class name and description of purpose be included on the
      same "printed page" as the copyright notice for easier
      identification within third-party archives.

   Copyright [yyyy] [name of copyright owner]

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
",
  "README.md": "# replace this",
  "package.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "devDependencies": {
      "@types/node": "*",
      "constructs": "^10.0.0",
      "projen": "*",
      "ts-node": "*",
      "typescript": "*",
    },
    "jest": {
      "projects": [],
    },
    "license": "Apache-2.0",
    "main": "lib/index.js",
    "name": "monorepo",
    "private": true,
    "publishConfig": {
      "access": "public",
    },
    "scripts": {
      "build": "npx projen build",
      "clobber": "npx projen clobber",
      "compile": "npx projen compile",
      "default": "npx projen default",
      "package": "npx projen package",
      "post-upgrade": "npx projen post-upgrade",
      "projen": "npx projen",
      "run": "npx projen run",
      "test": "npx projen test",
      "upgrade": "npx projen upgrade",
    },
    "types": "lib/index.d.ts",
    "version": "0.0.0",
    "workspaces": {
      "packages": [],
    },
  },
  "tsconfig.dev.json": {
    "compilerOptions": {
      "alwaysStrict": true,
      "declaration": true,
      "esModuleInterop": true,
      "experimentalDecorators": true,
      "inlineSourceMap": true,
      "inlineSources": true,
      "lib": [
        "es2019",
      ],
      "module": "CommonJS",
      "noEmitOnError": false,
      "noFallthroughCasesInSwitch": true,
      "noImplicitAny": true,
      "noImplicitReturns": true,
      "noImplicitThis": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "resolveJsonModule": true,
      "strict": true,
      "strictNullChecks": true,
      "strictPropertyInitialization": true,
      "stripInternal": true,
      "target": "ES2019",
    },
    "exclude": [
      "node_modules",
    ],
    "include": [
      ".projenrc.ts",
      "projenrc/**/*.ts",
    ],
    "references": [],
  },
  "tsconfig.json": {
    "compilerOptions": {
      "alwaysStrict": true,
      "declaration": true,
      "esModuleInterop": true,
      "experimentalDecorators": true,
      "inlineSourceMap": true,
      "inlineSources": true,
      "lib": [
        "es2019",
      ],
      "module": "CommonJS",
      "noEmitOnError": false,
      "noFallthroughCasesInSwitch": true,
      "noImplicitAny": true,
      "noImplicitReturns": true,
      "noImplicitThis": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "outDir": "lib",
      "resolveJsonModule": true,
      "rootDir": "src",
      "strict": true,
      "strictNullChecks": true,
      "strictPropertyInitialization": true,
      "stripInternal": true,
      "target": "ES2019",
    },
    "exclude": [],
    "include": [],
    "references": [],
  },
}
`;
